File: child_app/lib/models/goal.dart
class Goal {
  final String id;
  final String name;
  final String description;
  final double targetAmount;
  final double currentAmount;
  final String icon;
  final String color;
  final String category;
  final int priority;
  final bool isVisible;
  final bool isParallel;
  final String status;
  final DateTime? targetDate;
  final DateTime createdAt;
  final DateTime updatedAt;

  Goal({
    required this.id,
    required this.name,
    required this.description,
    required this.targetAmount,
    required this.currentAmount,
    required this.icon,
    required this.color,
    required this.category,
    required this.priority,
    required this.isVisible,
    required this.isParallel,
    required this.status,
    this.targetDate,
    required this.createdAt,
    required this.updatedAt,
  });

  // Goal sınıfında name yerine title kullanmak için getter ekleyelim
  String get title => name;

  double get progress => currentAmount / targetAmount;
  double get progressPercentage => (currentAmount / targetAmount) * 100;
  double get remainingAmount => targetAmount - currentAmount;
  bool get isCompleted => status == 'completed';
  bool get isActive => status == 'active';

  factory Goal.fromMap(Map<String, dynamic> map) {
    return Goal(
      id: map['_id']?.toString() ?? map['id']?.toString() ?? '',
      name: map['name'] ?? '',
      description: map['description'] ?? '',
      targetAmount: (map['targetAmount'] ?? 0).toDouble(),
      currentAmount: (map['currentAmount'] ?? 0).toDouble(),
      icon: map['icon'] ?? 'savings',
      color: map['color'] ?? '#2196F3',
      category: map['category'] ?? 'other',
      priority: map['priority'] ?? 1,
      isVisible: map['isVisible'] ?? true,
      isParallel: map['isParallel'] ?? false,
      status: map['status'] ?? 'active',
      targetDate:
          map['targetDate'] != null ? DateTime.parse(map['targetDate']) : null,
      createdAt: map['createdAt'] != null
          ? DateTime.parse(map['createdAt'])
          : DateTime.now(),
      updatedAt: map['updatedAt'] != null
          ? DateTime.parse(map['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'targetAmount': targetAmount,
      'currentAmount': currentAmount,
      'icon': icon,
      'color': color,
      'category': category,
      'priority': priority,
      'isVisible': isVisible,
      'isParallel': isParallel,
      'status': status,
      'targetDate': targetDate?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  Goal copyWith({
    String? name,
    String? description,
    double? targetAmount,
    double? currentAmount,
    String? icon,
    String? color,
    String? category,
    int? priority,
    bool? isVisible,
    bool? isParallel,
    String? status,
    DateTime? targetDate,
  }) {
    return Goal(
      id: id,
      name: name ?? this.name,
      description: description ?? this.description,
      targetAmount: targetAmount ?? this.targetAmount,
      currentAmount: currentAmount ?? this.currentAmount,
      icon: icon ?? this.icon,
      color: color ?? this.color,
      category: category ?? this.category,
      priority: priority ?? this.priority,
      isVisible: isVisible ?? this.isVisible,
      isParallel: isParallel ?? this.isParallel,
      status: status ?? this.status,
      targetDate: targetDate ?? this.targetDate,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }
}

class GoalCategory {
  static const String toy = 'toy';
  static const String electronics = 'electronics';
  static const String clothes = 'clothes';
  static const String sport = 'sport';
  static const String education = 'education';
  static const String travel = 'travel';
  static const String other = 'other';

  static const Map<String, String> names = {
    toy: '🧸 Oyuncak',
    electronics: '💻 Elektronik',
    clothes: '👕 Kıyafet',
    sport: '⚽ Spor',
    education: '📚 Eğitim',
    travel: '✈️ Seyahat',
    other: '🎯 Diğer',
  };

  static const Map<String, String> icons = {
    toy: 'toys',
    electronics: 'computer',
    clothes: 'checkroom',
    sport: 'sports_soccer',
    education: 'school',
    travel: 'flight',
    other: 'star',
  };
}

File: backend/src/models/Goal.js
import mongoose from 'mongoose';

const goalSchema = new mongoose.Schema({
  owner: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  name: {
    type: String,
    required: true,
    trim: true,
  },
  targetAmount: {
    type: Number,
    required: true,
    min: 0,
  },
  currentAmount: {
    type: Number,
    default: 0,
    min: 0,
  },
  description: {
    type: String,
    trim: true,
  },
  icon: {
    type: String,
    required: true,
  },
  status: {
    type: String,
    enum: ['active', 'paused', 'completed'],
    default: 'active',
  },
  targetDate: {
    type: Date,
  },
  priority: {
    type: Number,
    default: 1,
    min: 1,
  },
  isVisible: {
    type: Boolean,
    default: true,
  },
  isParallel: {
    type: Boolean,
    default: false,
  },
  color: {
    type: String,
    default: '#2196F3',
  },
  category: {
    type: String,
    enum: ['toy', 'electronics', 'clothes', 'sport', 'education', 'travel', 'other', 'game', 'games', 'book', 'books', 'money', 'car', 'house', 'phone', 'food', 'art', 'music'],
    default: 'other',
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
});

// İlerleme yüzdesi hesaplama
goalSchema.virtual('progress').get(function() {
  return (this.currentAmount / this.targetAmount) * 100;
});

// Hedef oluşturma statik metodu
goalSchema.statics.createGoal = async function (data) {
  const goal = new this(data);
  await goal.save();
  return goal;
};

// Para ekleme metodu
goalSchema.methods.addAmount = async function (amount) {
  if (this.status !== 'active') {
    throw new Error('Hedef aktif değil');
  }
  
  this.currentAmount += amount;
  this.updatedAt = new Date();
  
  // Hedef tamamlandı mı kontrol et
  if (this.currentAmount >= this.targetAmount) {
    this.status = 'completed';
    this.completedAt = new Date();
    this.currentAmount = this.targetAmount; // Hedef miktarını aşmayı engelle
  }
  
  // Kilometre taşlarını kontrol et
  for (const milestone of this.milestones) {
    if (!milestone.completed && this.currentAmount >= milestone.amount) {
      milestone.completed = true;
      milestone.completedAt = new Date();
    }
  }
  
  await this.save();
  return this;
};

// Hedef durumunu güncelleme metodu
goalSchema.methods.updateStatus = async function (status) {
  if (!['active', 'completed', 'paused'].includes(status)) {
    throw new Error('Geçersiz durum');
  }
  
  this.status = status;
  this.updatedAt = new Date();
  
  if (status === 'completed') {
    this.completedAt = new Date();
  }
  
  await this.save();
  return this;
};

const Goal = mongoose.model('Goal', goalSchema);

export { Goal }; 